### ISSUE TITLE
about mIoU metric

### ISSUE THREAD

**Issue by bbnwang** [RobvanGastel/dinov2-finetune#9]
https://github.com/RobvanGastel/dinov3-finetune/blob/fffecb4a1b2715550219bf98204c2f5af874e9e4/dino_finetune/metrics.py#L6

mIoU is not calculated in a right way. 

I follow the code as [mmsegmentation](https://github.com/open-mmlab/mmsegmentation/blob/main/mmseg/evaluation/metrics/iou_metric.py), got mIoU=52.09 on ADE20k with dinov3_vitl16.

**Comment by Fly-Yun-Maker** [comment]
Hello, I have been training the model for many days and found that the performance on the VOC dataset was unexpectedly high, reaching over 90%. I later discovered that this was due to an error in the mIoU  evaluation code. Could you please update it in time?

**Comment by RobvanGastel** [comment]
Sorry for the mistake, tonight I will have time to fix it. I will give you an update once finished

**Comment by RobvanGastel** [comment]
I found the mistake, I now first loop over the classes to then compute the IoU correctly.

```python
def compute_iou_metric(
    y_hat: torch.Tensor,
    y: torch.Tensor,
    ignore_index: Optional[int] = None,
    eps: float = 1e-6,
) -> float:
    """Compute the Intersection over Union metric for the predictions and labels.

    Args:
        y_hat (torch.Tensor): The prediction of dimensions (B, C, H, W), C being
            equal to the number of classes.
        y (torch.Tensor): The label for the prediction of dimensions (B, H, W)
        ignore_index (int | None, optional): ignore label to omit predictions in
            given region.
        eps (float, optional): To smooth the division and prevent division
        by zero. Defaults to 1e-6.

    Returns:
        float: The mean IoU
    """
    num_classes = int(y.max().item() + 1)
    y_hat = torch.argmax(y_hat, dim=1)
    
    ious = []
    for c in range(num_classes):
        y_hat_c = (y_hat == c)
        y_c = (y == c)

        # Ignore all regions with ignore
        if ignore_index is not None:
            mask = (y != ignore_index)
            y_hat_c = y_hat_c & mask
            y_c = y_c & mask

        intersection = (y_hat_c & y_c).sum().float()
        union = (y_hat_c | y_c).sum().float()

        if union > 0:
            ious.append((intersection + eps) / (union + eps))

    return torch.mean(torch.stack(ious)).item()
```

```python
def make_logits(pred, num_classes):
    b,h,w = pred.shape
    logits = torch.zeros((b, num_classes, h, w), dtype=torch.float32)
    for c in range(num_classes):
        logits[:, c] = (pred == c).float() * 10.0
    return logits


def test_miou_multiclass():
    """
    mIoU = (1 + 1/2 + 0)/3 = 2/3.
    """
    y = torch.tensor([[[0,1],[1,2]]], dtype=torch.long)
    y_hat = make_logits(torch.tensor([[[0,2],[1,1]]], dtype=torch.long), num_classes=3)

    expected = 4.0/9.0
    result = compute_iou_metric(y_hat, y)
    assert result == pytest.approx(expected, rel=1e-6)


def test_miou_with_ignore_index():
    """
    mIoU = (1 + 1/3 + 0)/3 = 4/9.
    """
    y = torch.tensor([[[0,0],[0,-1]]], dtype=torch.long)
    y_hat = make_logits(torch.tensor([[[0,0],[1,0]]], dtype=torch.long), num_classes=2)

    expected = 2.0/3.0
    result = compute_iou_metric(y_hat, y, ignore_index=-1)
    assert result == pytest.approx(expected, rel=1e-6)


test_miou_ignore_index_masks_not_relabel()
test_miou_multiclass()
```

I ran these tests and with the new IoU calculation, I get the correct results now. I will rerun the metrics and also update the README.md.
